Feat: Whitelist tokenfactory hooks

jcompagni10 opened the pull request:
Add explicit whitelist of codeID & denomCreator pairs to tokenfactory params. Only allow beforeSend hooks to be called for denoms where the denomCreator and codeID of hook are whitelisted.

This also adds a migration for the tokefactory params that includes hooks that we want to whitelist (NFA and astroport)

**related PRs:**
https://github.com/neutron-org/neutron-integration-tests/pull/316 --test fixes for this PR to pass

neutron dao changes: 
https://github.com/neutron-org/neutron-dao/pull/107
https://github.com/neutron-org/neutron-sdk/pull/150
https://github.com/neutron-org/neutronjsplus/pull/43
https://github.com/neutron-org/neutron-integration-tests/pull/314

mmikev commented on proto/osmosis/tokenfactory/v1beta1/genesis.proto:
```diff
@@ -29,4 +29,5 @@ message GenesisDenom {
     (gogoproto.moretags) = "yaml:\"authority_metadata\"",
     (gogoproto.nullable) = false
   ];
+  bool whitelisted = 3 [(gogoproto.moretags) = "yaml:\"whitelisted\""];
```
Can't we just store whitelisted denoms in module's params literally as a list of tokens?

It's more convenient for maintenance

jcompagni10 committed:
basic whitelist logic

jcompagni10 committed:
Fix tests

jcompagni10 committed:
Add note on weird TF code

jcompagni10 committed:
Block adding non-whitelisted hooks

jcompagni10 committed:
Refactor + add more tests

jcompagni10 committed:
add migration

jcompagni10 committed:
lint

jcompagni10 committed:
Merge remote-tracking branch 'origin/main' into feat/whitelist-tf-hooks

sotnikov-s commented on proto/osmosis/tokenfactory/params.proto:
```diff
@@ -0,0 +1,39 @@
+
+option go_package = "github.com/neutron-org/neutron/v4/x/tokenfactory/types";
+
+message HookWhitelist {
```
maybe WhitelistedHook? HookWhitelist sounds like it's the list itself, not an entry in it

sotnikov-s replied:
please add a descriptive comment to the type

sotnikov-s commented on proto/osmosis/tokenfactory/params.proto:
```diff
@@ -0,0 +1,39 @@
+  // FeeCollectorAddress is the address where fees collected from denom creation
+  // are sent to
+  string fee_collector_address = 3;
+  repeated HookWhitelist whitelisted_hooks = 4;
```
please add a descriptive comment to the field

sotnikov-s commented on x/tokenfactory/module.go:
```diff
@@ -135,6 +135,10 @@ func (AppModule) QuerierRoute() string { return types.QuerierRoute }
 	types.RegisterMsgServer(cfg.MsgServer(), keeper.NewMsgServerImpl(am.keeper))
+	m := keeper.NewMigrator(am.keeper)
+	if err := cfg.RegisterMigration(types.ModuleName, 1, m.Migrate1to2); err != nil {
+		panic(fmt.Sprintf("failed to migrate x/tokenfactory from version 1 to 1: %v", err))
```
```suggestion
		panic(fmt.Sprintf("failed to migrate x/tokenfactory from version 1 to 2: %v", err))
```

zavgorodnii replied:
`version 1 to 1` doesn't look right

sotnikov-s commented on x/tokenfactory/keeper/before_send.go:
```diff
@@ -99,6 +99,11 @@ func (k Keeper) callBeforeSendListener(ctx context.Context, from, to sdk.AccAddr
+			if err := k.AssertIsHookWhitelisted(c, coin.Denom, cwAddr); err != nil {
+				return nil
+			}
+
```
once have been discussed and based on what we decide will be the best behaviour here, please update the `callBeforeSendListener` to reflect the change

jcompagni10 replied:
Just confirming we are good with this behavior? @zavgorodnii @sotnikov-s

sotnikov-s replied:
even though we remove all nonwhitelisted hooks during migration, I still think it's more appropriate to return an error here. we have two options here in case of an unauthorised hook: 1) we completely ignore all the hooks code (return nil here breaking the method execution) or 2) we return error here explicitly making all bank operations with unauthorised hooks fail. I prefer the second one

sotnikov-s replied:
if we still stick to the first option described above, we should replace `return nil` with `continue` because the `return nil` happens in a loop of handling a bank operation that possibly contains multiple assets. if just return, we break the loop and ignore the rest of the assets

jcompagni10 replied:
Spoke with Andrei about this. This line is really only here as an emergency override in case something is wrong with a hook and we want to use the sub-dao to turn it off immediately. This is an extreme case and just a stop-gap before we would do an emergency upgrade to remove the problematic hook. In this scenario we do not want the offending hook to be able to block sends. 
Agree on switching to `continue` though.

sotnikov-s replied:
please also add an error logging here

sotnikov-s commented on x/tokenfactory/types/params.go:
```diff
@@ -105,3 +113,24 @@ func validateFeeCollectorAddress(i interface{}) error {
+	for _, hook := range hooks {
+		hookStr := hook.String()
+		if seenHooks[hookStr] {
+			return fmt.Errorf("duplicate whitelisted hook: %#v", hook)
```
what does `%#v` do? why not `%s` with `hookStr` as value?

jcompagni10 replied:
It is supposed to pretty-print it. But actually comparing the output I agree, hookStr is much clearer

sotnikov-s commented on x/tokenfactory/types/tx.go:
```diff
@@ -32,9 +32,11 @@ func (msg *MsgUpdateParams) Validate() error {
 		return errorsmod.Wrap(err, "authority is invalid")
 	}
 
+	// TODO: This is inconsistent. Per Params.Validate() an empty creator address is valid as long as
+	// DenomCreationFee is nil. But This check fails if FeeCollectorAddress is unset.
 	if _, err := sdk.AccAddressFromBech32(msg.Params.FeeCollectorAddress); err != nil {
 		return errorsmod.Wrap(err, "fee_collector_address is invalid")
 	}
```
why do we even need this piece of code? why there is a part of `msg.Params` validation placed here, not in the `msg.Params.Validate`? we literally have the same validation inside the `msg.Params.Validate` method:
<img width="605" alt="image" src="https://github.com/neutron-org/neutron/assets/34917380/1da66de6-1e15-4ad2-bf75-c46035de7823">

jcompagni10 replied:
ValidateFeeCollectorAddress allows an empty address. But params.Validate() does not. I'm fairly certain we should remove the validation of FeeCollectorAddress in params.Validate(). But doing so breaks some tests. So in the interest of keeping the commit as clear and atomic as possible, I did not want to change anything that wasn't essential.

jcompagni10 committed:
add migration to remove non-whitelisted tf hooks

jcompagni10 committed:
lint

jcompagni10 committed:
misc review fixes

jcompagni10 committed:
regen swagger

jcompagni10 commented:
https://github.com/neutron-org/neutron-tests/actions/runs/9576807194

sotnikov-s commented on x/tokenfactory/migrations/v2/store.go:
```diff
@@ -0,0 +1,112 @@
+				store.Delete(iterator.Key())
+			}
+		}
+	}
```
a couple of concerns about this piece of code:
1. there are different kinds of errors returned from the `AssertIsHookWhitelisted`, but here they all are treated as the "hook is not whitelisted" case. there are also possible errors about invalid denoms and not found contracts. this is an imperfection of having the only return value that is used for both error values and true/false value. I think it would be cleaner to refactor `AssertIsHookWhitelisted` so it returns `(bool, error)`
2. the `keyParts[2]` and `keyParts[1]` — we never check whether the array is long enough beforehand which can cause to panic. it's a good practice to check the length before referring to array elements using index

I understand that it's a migration code where we kind of know the values we'll operate with and we can be sure about the array index operations and the only possible error returned from `AssertIsHookWhitelisted` (because we're sure the denom and contract address are fine), but I still find the code unclean and rough here. so I'd appreciate to see the code refactored in case you share my concerns and we have some time for doing that.
besides, if we're so sure about the values, why we handle the `AccAddressFromBech32` error? that's inconsistent

jcompagni10 replied:
Reposting from slack for others context:

So thinking about it more I'm not sure I totally agree with this. After the migration when we use this check in MsgServer it doesn't really matter why the hook isn't whitelisted (invalid contract, invalid denom, no whitelist value) we are always going to return an error. So we might as well use AssertIsHookWhitelisted to generate the errors so they are as informative as possible. If it returns false, nil for the case where there are no "errors" but the hook is not whitelisted then the message server  code has to be
```
whitelisted, err := server.Keeper.AssertIsHookWhitelisted(ctx, msg.Denom, cwAddr)
		if err != nil {
			return nil, err
		}
		if !whitelisted {
			return nil, types.ErrBeforeSendHookNotWhitelisted
		}
```
Which feels messy. Also, we no longer have the context to return an informative error message like we do here:  https://github.com/neutron-org/neutron/blob/a6eae66545df6e128da2c7cac81ff4c1c02deb7b/x/tokenfactory/keeper/params.go#L51
In the case of the migration, I agree that it's not quite perfect. but given that we are migrating a small set of specific hooks I think we can fairly assume that those are correct. If there are other hooks that throw an error as part of the check we do not want to stop the migration. For example, if someone added a hook before the upgrade the pointed to a non-existent contract we wouldn't want the migration to fail.
Let me know what you think?


I will deal with the array case better.

sotnikov-s replied:
okay, I see. what do you think about the following approach then?
1. make the errors returned from `AssertIsHookWhitelisted` more specific in terms of types:
<img width="1420" alt="image" src="https://github.com/neutron-org/neutron/assets/34917380/ee946213-2872-4c54-ae67-8a6360dd755c">


2. remove hooks only in case of the `types.ErrBeforeSendHookNotWhitelisted error`, otherwise fail:
<img width="1426" alt="image" src="https://github.com/neutron-org/neutron/assets/34917380/e47d04e8-7d80-4ee9-8d28-b49e8a95a527">

I think it's a quite good enough alternative here

jcompagni10 replied:
The solution you describe above would fail for any hooks where the current contract is not valid. I believe there is at least one such contract right now.

sotnikov-s commented on x/tokenfactory/keeper/params.go:
```diff
@@ -29,3 +29,24 @@ func (k Keeper) SetParams(ctx sdk.Context, params types.Params) error {
 	return nil
 }
+
+func (k Keeper) AssertIsHookWhitelisted(ctx sdk.Context, denom string, contractAddress sdk.AccAddress) error {
```
please move the method to `keeper.go` or `before_send.go`, or to any other file you find is the best place for this method, because `params.go` is definitely not the one

jcompagni10 committed:
PR review fixes

zavgorodnii commented on docs/static/swagger.yaml:
```diff
@@ -18774,6 +18795,19 @@ definitions:
+                  type: string
+              type: object
+            title: >-
+              HookWhitelist is the list of hooks which are allowed to be added
```
might be missing something, but why do we have `whitelisted_hooks` as the key here and `HookWhitelist` as the value?

jcompagni10 replied:
ah need to regen swagger

jcompagni10 committed:
Merge remote-tracking branch 'origin/main' into feat/whitelist-tf-hooks

jcompagni10 committed:
small migration fixes

jcompagni10 committed:
fix: dont write during iterations

jcompagni10 committed:
lint

jcompagni10 committed:
Add error logging for skipped hooks

jcompagni10 committed:
Add whitelisted astroport contracts to TF migration

jcompagni10 committed:
Add missing comma in init-neutrond.sh

jcompagni10 committed:
better comments on TF params.proto

jcompagni10 commented:
testrun with all PRs: https://github.com/neutron-org/neutron-tests/actions/runs/9606960942

sotnikov-s commented on x/tokenfactory/migrations/v2/store.go:
```diff
@@ -0,0 +1,128 @@
+	for ; iterator.Valid(); iterator.Next() {
+		keyParts := strings.Split(string(iterator.Key()), types.KeySeparator)
+		if len(keyParts) != 3 {
+			return errors.New("cannot parse BeforeSendHook data")
```
please add the keyparts print here so the error message is comprehensive

sotnikov-s commented on x/tokenfactory/keeper/before_send.go:
```diff
@@ -99,6 +120,20 @@ func (k Keeper) callBeforeSendListener(ctx context.Context, from, to sdk.AccAddr
+				c.Logger().Error(
+					"Skipped hook execution due to missing whitelist",
+					"err", err,
+					"denom", coin.Amount,
```
```suggestion
					"denom", coin.Denom,
```

sotnikov-s commented on x/tokenfactory/migrations/v2/store.go:
```diff
@@ -0,0 +1,128 @@
+			denom := keyParts[1]
+			contractAddr, err := sdk.AccAddressFromBech32(string(iterator.Value()))
+			if err != nil {
+				return errors.New("cannot parse hook contract address")
```
please add the contractAddr print here so the error message is comprehensive

sotnikov-s commented on x/tokenfactory/types/params.go:
```diff
@@ -54,6 +57,10 @@ func (p Params) Validate() error {
 	}
 
+	if err := validateWhitelistedHooks(p.WhitelistedHooks); err != nil {
+		return fmt.Errorf("failed to validate params: %w", err)
```
```suggestion
		return fmt.Errorf("failed to validate whitelisted hooks: %w", err)
```

pr0n00gler committed:
Merge branch 'main' into feat/whitelist-tf-hooks

jcompagni10 committed:
make errors clearer

jcompagni10 committed:
fixes to tokenfactory query cli

1. Add missing GetCmdBeforeSendHook query
2. Correctly parse denoms with slashes in subdenom
3. fix typo hiik => hook

jcompagni10 committed:
update code_ids for astroport contract whitelisting

jcompagni10 commented:
Testrun with just neutron changes:https://github.com/neutron-org/neutron-tests/actions/runs/9608462857

pr0n00gler committed:
do not use a separate event manager for sudo calls in tokenfactory
